pkg/egressgateway/policy.go

```diff
@@ -55,32 +62,37 @@ type PolicyConfig struct {
 
 	policyGwConfig *policyGatewayConfig
 
-	matchedEndpoints map[endpointID]*endpointMetadata
+	matchedEndpoints []*k8sTypes.CiliumEndpoint
 	gatewayConfig    gatewayConfig
 }
 
 // PolicyID includes policy name and namespace
 type policyID = types.NamespacedName
 
-// matchesEndpointLabels determines if the given endpoint is a match for the
+// matchesEndpointLabels determines if the given endpoint labels match for the
 // policy config based on matching labels.
-func (config *PolicyConfig) matchesEndpointLabels(endpointInfo *endpointMetadata) bool {
-	labelsToMatch := k8sLabels.Set(endpointInfo.labels)
+func (config *PolicyConfig) matchesEndpointLabels(epLabels k8sLabels.Labels) bool {
 	for _, selector := range config.endpointSelectors {
-		if selector.Matches(labelsToMatch) {
+		if selector.Matches(epLabels) {
 			return true
 		}
 	}
 	return false
 }
 
-// updateMatchedEndpointIDs update the policy's cache of matched endpoint IDs
-func (config *PolicyConfig) updateMatchedEndpointIDs(epDataStore map[endpointID]*endpointMetadata) {
-	config.matchedEndpoints = make(map[endpointID]*endpointMetadata)
+// updateMatchedEndpoints update the policy's cache of matched endpoint IDs
+func (config *PolicyConfig) updateMatchedEndpoints(epStore resource.Store[*k8sTypes.CiliumEndpoint], idAllocator identityCache.IdentityAllocator) {
+	config.matchedEndpoints = nil
 
-	for _, endpoint := range epDataStore {
-		if config.matchesEndpointLabels(endpoint) {
-			config.matchedEndpoints[endpoint.id] = endpoint
+	for _, endpoint := range epStore.List() {
+		epLabels, err := getEndpointLabels(idAllocator, endpoint)
```

Comment:
This seems dangerously inefficient. We're iterating potentially very often over all endpoints (there may be thousands of them) and doing point-wise lookups to the identity allocator, a core piece of the agent. Increased lock contention there will slow everything down.

Reply:
CC @jibi 
Thanks to 1s `EgressGatewayReconciliationTriggerInterval` maybe this is manageable for the time being until a better solution is implemented?

Reply:
I initially missed the fact that we are fetching _all_ endpoints' labels each time we run the reconciliation :disappointed:

I don't know if the 1s ratelimit will be enough or not (as this also depends on the number of endpoints in the cluster), I'll have to measure some numbers.

And I guess having a local cache for labels (that gets updated only once per endpoint event in `handleEndpointEvent()`) defeats the purpose of the new `epStore` as the 2 can go out of sync, e.g. we can iterate over an endpoint in `epStore` for which we didn't receive yet the event and so labels are not yet updated?